File: backend/app/models/instance.py
"""
实例数据模型
定义 MaiBot 机器人实例的数据结构
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class InstanceStatus(str, Enum):
    """实例状态枚举"""
    STOPPED = "stopped"
    RUNNING = "running"
    STARTING = "starting"
    STOPPING = "stopping"
    ERROR = "error"


class BotType(str, Enum):
    """机器人类型"""
    MAIBOT = "maibot"
    NAPCAT = "napcat"
    OTHER = "other"


class InstanceBase(BaseModel):
    """实例基础信息"""
    name: str = Field(..., min_length=1, max_length=100, description="实例名称")
    bot_type: BotType = Field(default=BotType.MAIBOT, description="机器人类型")
    bot_version: Optional[str] = Field(None, description="机器人版本")
    description: Optional[str] = Field(None, max_length=500, description="实例描述")


class InstanceCreate(InstanceBase):
    """创建实例请求模型"""
    python_path: Optional[str] = Field(None, description="Python 路径")
    config_path: Optional[str] = Field(None, description="配置文件路径")


class InstanceUpdate(BaseModel):
    """更新实例请求模型"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    python_path: Optional[str] = None
    config_path: Optional[str] = None


class Instance(InstanceBase):
    """实例完整信息"""
    id: str = Field(..., description="实例唯一标识符")
    instance_path: Optional[str] = Field(None, description="实例目录路径")
    status: InstanceStatus = Field(default=InstanceStatus.STOPPED, description="实例状态")
    python_path: Optional[str] = None
    config_path: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now, description="创建时间")
    updated_at: datetime = Field(default_factory=datetime.now, description="更新时间")
    last_run: Optional[datetime] = Field(None, description="最后运行时间")
    run_time: int = Field(default=0, description="总运行时间（秒）")
    
    class Config:
        json_schema_extra = {
            "example": {
                "id": "inst_1234567890",
                "name": "MaiBot 生产实例",
                "bot_type": "maibot",
                "bot_version": "1.0.0",
                "description": "主要的机器人实例",
                "status": "stopped",
                "python_path": "/usr/bin/python3",
                "config_path": "/path/to/config.json",
                "created_at": "2024-01-01T00:00:00",
                "updated_at": "2024-01-01T00:00:00",
            }
        }


class InstanceStatusResponse(BaseModel):
    """实例状态响应模型"""
    id: str
    status: InstanceStatus
    pid: Optional[int] = Field(None, description="进程 ID")
    uptime: Optional[int] = Field(None, description="运行时间（秒）")


class InstanceList(BaseModel):
    """实例列表响应模型"""
    total: int = Field(..., description="总实例数")
    instances: List[Instance] = Field(..., description="实例列表")

File: frontend/src-tauri/src/models/instance.rs
/// 实例数据模型
///
/// 对应 Python 的 InstanceDB 表，以及前端 instanceApi.ts 中的接口定义。
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};

/// 机器人实例（对应 instances 表）
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Instance {
    /// 实例唯一标识符
    pub id: String,
    /// 实例名称
    pub name: String,
    /// 实例目录路径（相对于 instances_dir）
    pub instance_path: Option<String>,
    /// 机器人类型（maibot / napcat / other）
    pub bot_type: String,
    /// 机器人版本
    pub bot_version: Option<String>,
    /// 实例描述
    pub description: Option<String>,
    /// 实例状态（stopped / running / error 等）
    pub status: String,
    /// Python 路径
    pub python_path: Option<String>,
    /// 配置文件路径
    pub config_path: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    /// 最后运行时间
    pub last_run: Option<NaiveDateTime>,
    /// 总运行时间（秒）
    pub run_time: i64,
    /// QQ 账号（NapCat 快速登录）
    pub qq_account: Option<String>,
    /// CPU 使用率（运行时由进程管理器填充，不存储于数据库）
    #[sqlx(skip)]
    pub cpu_usage: Option<f64>,
    /// 内存使用量 MB（运行时由进程管理器填充，不存储于数据库）
    #[sqlx(skip)]
    pub memory_usage: Option<f64>,
}

/// 创建实例请求参数（对应前端 InstanceCreate 接口）
#[derive(Debug, Deserialize)]
pub struct CreateInstanceRequest {
    pub name: String,
    pub bot_type: Option<String>,
    pub bot_version: Option<String>,
    pub description: Option<String>,
    pub python_path: Option<String>,
    pub config_path: Option<String>,
}

/// 更新实例请求参数（对应前端 InstanceUpdate 接口，全部字段可选）
#[derive(Debug, Deserialize)]
pub struct UpdateInstanceRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub python_path: Option<String>,
    pub config_path: Option<String>,
    pub qq_account: Option<String>,
}

/// 实例列表响应（对应前端 InstanceList 接口）
#[derive(Debug, Serialize)]
pub struct InstanceList {
    pub total: usize,
    pub instances: Vec<Instance>,
}

/// 实例运行状态响应（对应前端 InstanceStatusResponse 接口）
#[derive(Debug, Serialize)]
pub struct InstanceStatusResponse {
    pub id: String,
    pub status: String,
    pub pid: Option<u32>,
    pub uptime: Option<f64>,
}

/// 组件运行状态（对应前端 ComponentStatus 接口）
#[derive(Debug, Serialize)]
pub struct ComponentStatus {
    pub component: String,
    pub running: bool,
    pub pid: Option<u32>,
    pub uptime: Option<f64>,
}

/// 通用成功响应（对应前端 SuccessResponse 接口）
#[derive(Debug, Serialize)]
pub struct SuccessResponse {
    pub success: bool,
    pub message: String,
}

impl SuccessResponse {
    pub fn ok(message: impl Into<String>) -> Self {
        Self {
            success: true,
            message: message.into(),
        }
    }
}
